### PR TITLE
KAFKA-3973: feasibility caching bytes vs records

### PR DESCRIPTION
**Pull request by bbejeck**
@enothereska 

Added the benchmark test per comments.

Thanks

EDIT: Results from the test and additional comments are in the Jira ticket for KAFKA-3973

### CODE REVIEW

**Comment by bbejeck** [comment]
I checked the output from the build and it looks like the failure is not related to this PR:

> org.apache.kafka.common.network.SslSelectorTest > testRenegotiation STARTED
> Build timed out (after 180 minutes). Marking the build as failed.
> Build was aborted

File: build.gradle [hunk 1]
```diff
@@ -690,6 +690,7 @@ project(':streams') {
     testCompile project(':core')
     testCompile project(':core').sourceSets.test.output
     testCompile libs.junit
+    testCompile 'com.github.jbellis:jamm:0.3.1'
```
Comment (enothereska) [inline build.gradle:693]: @ijuma are you ok with this addition?

File: streams/src/test/java/org/apache/kafka/streams/perf/CachingPerformanceBenchmarks.java [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+import org.apache.kafka.test.MockMemoryLRUCache;
+
+
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
```
Comment (enothereska) [inline streams/src/test/java/org/apache/kafka/streams/perf/CachingPerformanceBenchmarks.java]: Perhaps a top-level comment on what this class does would be good.

**Comment by bbejeck** [comment]
@enothereska 

Comments describing class actions added.

**Comment by bbejeck** [comment]
ping @ijuma  @guozhangwang

File: build.gradle (hunk 1 above)
Comment (ijuma) [inline build.gradle:693]: Have we decided that we will use jamm? Or is this exploratory still? If the latter, is there benefit in merging the code at this stage?

Comment (bbejeck) [inline build.gradle:693]: Still exploratory as far as I know.  The only value in merging is if we want to merge in the benchmarking code for documentation purposes.

**Comment by enothereska** [comment]
LGTM. @ijuma @guozhangwang Although this is exploration as part of KIP-63 I don't see a reason why the benchmark cannot be committed as is, since it's useful beyond KIP-63.

File: streams/src/test/java/org/apache/kafka/streams/perf/CachingPerformanceBenchmarks.java [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+            @Override
+            public void apply(K key, V value) {
+                throw new IllegalStateException("Should be no evictions");
+            }
+        };
+    }
+
+    private static class BlackHole<T> {
+        public  void takeValue(T value) {
+            value = null;
```
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/perf/CachingPerformanceBenchmarks.java:310]: The JIT can easily inline this method, so it doesn't actually do anything. The BlackHole implementation in JMH is a lot more complex: http://hg.openjdk.java.net/code-tools/jmh/file/cde312963a3d/jmh-core/src/main/java/org/openjdk/jmh/logic/BlackHole.java#l117

**Comment by ijuma** [comment]
@enothereska It's really hard to write correct micro-benchmarks in the JVM by hand. That's why I strongly recommend JMH. So, if we want to merge this, then we need to review it to check that it is measuring what it claims to be measuring (I saw a potential problem in my very quick analysis).

It seems that @bbejeck was interested in contributing a JMH-based module for Kafka, I think that would be a better route, personally.

**Comment by mjsax** [comment]
Closing this old PR -- the corresponding Jira is marked as "resolved".

**Pull request closed without merging**